### PR TITLE
Use gogetdoc instead of godef and godoc

### PR DESCRIPTION
**Pull request by ramya-rao-a**
Currently we use `godef` for Goto/Peek Definition.
`godef` + `godoc` together provide the Hover Info and Signature Help.

These tools have shortcomings which we hope to overcome by using `gogetdoc` instead.
#440 #442 #515 #567 #459 #496

This PR is work on top of what @leaxoy has done in #607 
What has been added are the below
- A setting where user can choose `godoc` (old way of doing things `godef` + `godoc`)  vs `gogetdoc`
- If User is using Go 1.5, then we use the old way regardless of what is set
- The setting will default to the old way. There are some 3 or 4 issues logged by users which using `gogetdoc` will solve. We will reach out to those users and ask them to change the setting to use `gogetdoc` and give feedback
- Some refactoring
- Fix the failing  tests run by travis
- Add unit tests to cover the scenarios mentioned in the above mentioned issues

### CODE REVIEW

**Comment by ramya-rao-a** [comment]
the unit tests are failing for 1.7 but pass in 1.6 mac... which is baffling

@leaxoy if you have some time to look at the test failures, I'd appreciate it

File: test/go.test.ts [hunk 1]
```diff
@@ -53,28 +53,38 @@ encountered.
...
+			if (version.major > 1 || (version.major === 1 && version.minor > 5)) {
+				testCases[0][1] = 'func main()';
+				testCases[1][1] = 'package fmt';
+				testCases[2][1] = 'func Println(a ...interface{}) (n int, err error)';
+				testCases[3][1] = 'func print(txt string)';
+			}
+			return vscode.workspace.openTextDocument(uri).then((textDocument) => {
+				let promises = testCases.map(([position, expectedSignature, expectedDocumentation]) =>
+					provider.provideHover(textDocument, position, null).then(res => {
+						// TODO: Documentation appears to currently be broken on Go 1.7, so disabling these tests for now
```
Comment (abarisain) [inline test/go.test.ts]: Can we give reenabling these tests a shot? When this was commented, they worked on my machine

File: src/goSignature.ts [hunk 2]
```diff
@@ -28,10 +28,25 @@ export class GoSignatureHelpProvider implements SignatureHelpProvider {
...
+				let sigStart = nameEnd + 5; // ' func'
+				let funcName = text.substring(0, nameEnd);
+				sig = text.substring(sigStart);
+				si = new SignatureInformation(funcName + sig, res.doc);
+			} else {
+				// declaration is of the form "func Add(a int, b int) int"
+				text = res.declarationlines[0].substring(5);
+				si = new SignatureInformation(text, res.doc);
+				let braceStart = text.indexOf('(');
+				sig = text.substring(braceStart);
```
Comment (abarisain) [inline src/goSignature.ts]: Does that work for methods that have a signature like `func (l *Struct) Add(a int, b int) {`? I did it somewhat differently for that reason https://github.com/abarisain/vscode-go/blob/gogetdoc2/src/goSignature.ts#L33-L40

If it does, then sorry! Haven't had the time to test this branch yet.

**Comment by ramya-rao-a** [comment]
All tests are a pass!

File: src/goSignature.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/goSignature.ts]: Good catch, ya it doesnt work for that case

Comment (ramya-rao-a) [inline src/goSignature.ts]: In such case it thinks the receiver is the first param, thats a bug.

Comment (ramya-rao-a) [inline src/goSignature.ts]: Fixed it, Took your code and tested

Comment (ramya-rao-a) [inline src/goSignature.ts]: @abarisain The changes are in

File: test/go.test.ts (hunk 1 above)
Comment (ramya-rao-a) [inline test/go.test.ts]: @abarisain Done!

**Comment by ramya-rao-a** [comment]
Added additional tests to cover various scenarios where using `gogetdoc` wins vs `godef`

File: src/goInstallTools.ts [hunk 3]
```diff
@@ -22,13 +22,19 @@ function getTools(goVersion: SemVersion): { [key: string]: string } {
...
 		'go-outline': 'github.com/lukehoban/go-outline',
 		'go-symbols': 'github.com/newhook/go-symbols',
 		'guru': 'golang.org/x/tools/cmd/guru',
 		'gorename': 'golang.org/x/tools/cmd/gorename'
 	};
 
+	// Install the doc/def tool that was chosen by the user
+	if (goConfig['docsTool'] === 'godoc') {
+		tools['godef'] = 'github.com/rogpeppe/godef';
+	} else if (goConfig['formatTool'] === 'goreturns') {
```
Comment (abarisain) [inline src/goInstallTools.ts]: I think that should be `} else if (goConfig['docsTool'] === 'gogetdoc') {` :)

**Comment by zmb3** [comment]
LGTM :-)

**Comment by abarisain** [comment]
Ditto

**Comment by abarisain** [comment]
Great addition to the plugin ! Congrats @zmb3

**Pull request merged**